test/e2e/framework/framework.go

```diff
@@ -60,3 +66,122 @@ func queryAndUpdateGlobalnetStatus() {
+			}
+			return true, "", nil
+		})
+	return gw.(*submarinerv1.Gateway)
+}
+
+func (f *Framework) AwaitForGatewayRemoved(cluster framework.ClusterIndex, name string) {
+	gwClient := SubmarinerClients[cluster].SubmarinerV1().Gateways(framework.TestContext.SubmarinerNamespace)
+	framework.AwaitUntil(fmt.Sprintf("await Gateway on %q removed", name),
+		func() (interface{}, error) {
+			_, err := gwClient.Get(name, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, err
+		},
+		func(result interface{}) (bool, string, error) {
+			gone := result.(bool)
+			return gone, "", nil
+		})
+
+}
+
+func (f *Framework) AwaitGatewayFullyConnected(cluster framework.ClusterIndex, name string) *submarinerv1.Gateway {
+	gwClient := SubmarinerClients[cluster].SubmarinerV1().Gateways(framework.TestContext.SubmarinerNamespace)
+	gw := framework.AwaitUntil(fmt.Sprintf("await Gateway on %q with status active and connections UP", name),
+		func() (interface{}, error) {
+			resGw, err := gwClient.Get(name, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return nil, nil
+			}
+			return resGw, err
+		},
+		func(result interface{}) (bool, string, error) {
+			if result == nil {
+				return false, "gateway not found yet", nil
+			}
+			gw := result.(*submarinerv1.Gateway)
+			if gw.Status.HAStatus != submarinerv1.HAStatusActive {
+				return false, fmt.Sprintf("Gateway %q exists but not active yet",
+					gw.Name), nil
+			}
+			if len(gw.Status.Connections) == 0 {
+				return false, fmt.Sprintf("Gateway %q exist but has no connections yet", name), nil
+			}
+			for _, conn := range gw.Status.Connections {
+				if conn.Status != submarinerv1.Connected {
+					return false, fmt.Sprintf("Gateway %q exist but connection to cluster %q is not up yet",
+						name, conn.Endpoint.ClusterID), nil
+				}
+			}
+
+			return true, "", nil
+		})
+	return gw.(*submarinerv1.Gateway)
+}
+
+// GatewayCleanup ensures that only the active gateway node is flagged as gateway node
+//                which could not be after a failed test which left the system on an
+//                unexpected state
+func (f *Framework) GatewayCleanup() {
+
+	passiveGateways := f.GetGatewaysWithHAStatus(submarinerv1.HAStatusPassive)
+
+	if len(passiveGateways) == 0 {
+		return
+	}
+
+	ginkgo.By(fmt.Sprintf("Cleaning up any non-active gateways: %v", gatewayNames(passiveGateways)))
+	for _, nonActiveGw := range passiveGateways {
+		f.SetGatewayLabelOnNode(framework.ClusterA, nonActiveGw.Name, false)
+		f.AwaitForGatewayRemoved(framework.ClusterA, nonActiveGw.Name)
+	}
+}
+
+func gatewayNames(gateways []submarinerv1.Gateway) []string {
+	names := []string{}
+	for _, gw := range gateways {
+		names = append(names, gw.Name)
+	}
+	return names
+}
+
+func (f *Framework) GetGatewaysWithHAStatus(status submarinerv1.HAStatus) []submarinerv1.Gateway {
+
+	gatewayClient := SubmarinerClients[framework.ClusterA].SubmarinerV1().Gateways(
```

Comment:
This should really pass in the `ClusterIndex` instead of assuming `ClusterA` to be consistent with other such functions.

Reply:
oh true!

Reply:
Done.